aten/src/ATen/native/LossNLL2d.cpp

```diff
@@ -0,0 +1,410 @@
+template <typename scalar_t>
+inline scalar_t* optional_data(const Tensor& source) {
+  return source.defined() ? source.data_ptr<scalar_t>() : nullptr;
+}
+
+inline void check_inputs_nll_loss2d(
+    const Tensor& input,
+    const Tensor& target,
+    const Tensor& weight) {
+  TORCH_CHECK(
+      target.dim() == 3,
+      "only batches of spatial targets supported (3D tensors)"
+      " but got targets of dimension: ",
+      target.dim());
+  TORCH_CHECK(
+      input.dim() == 4,
+      "only batches of spatial inputs supported (4D tensors), "
+      "but got input of dimension: ",
+      input.dim());
+  TORCH_CHECK(
+      !weight.defined() || weight.numel() == input.size(1),
+      "weight tensor should be defined either for all or no classes");
+
+  const int64_t input0 = input.size(0);
+  const int64_t input2 = input.size(2);
+  const int64_t input3 = input.size(3);
+  const int64_t target0 = target.size(0);
+  const int64_t target1 = target.size(1);
+  const int64_t target2 = target.size(2);
+  TORCH_CHECK(
+      input0 == target0 && input2 == target1 && input3 == target2,
+      "size mismatch (got input: ",
+      input.sizes(),
+      " , target: ",
+      target.sizes());
+}
+
+inline void check_gradout_shape_nll_loss2d(
+    const Tensor& grad_output,
+    const Tensor& target) {
+  TORCH_CHECK(
+      grad_output.dim() == 3,
+      "grad_output must have same dimension as target (3) but got dimension: ",
+      grad_output.sizes());
+
+  const int64_t grad_output0 = grad_output.size(0);
+  const int64_t grad_output1 = grad_output.size(1);
+  const int64_t grad_output2 = grad_output.size(2);
+  const int64_t target0 = target.size(0);
+  const int64_t target1 = target.size(1);
+  const int64_t target2 = target.size(2);
+  TORCH_CHECK(
+      grad_output0 == target0 && grad_output1 == target1 &&
+          grad_output2 == target2,
+      "size mismatch (got grad_output: ",
+      grad_output.sizes(),
+      " target: ",
+      target.sizes());
+}
+
+template <typename scalar_t>
+static void nll_loss2d_forward_out_frame(
+    Tensor& output,
+    Tensor& total_weight,
+    const Tensor& input,
+    const Tensor& target,
+    const Tensor& weight,
+    int64_t reduction,
+    int64_t ignore_index) {
+  const int64_t n_classes = input.size(1);
+
+  scalar_t* total_weight_data = total_weight.data_ptr<scalar_t>();
+  *total_weight_data = 0;
+
+  auto weight_contiguous = optional_contiguous(weight);
+  const scalar_t* weight_data = optional_data<scalar_t>(weight_contiguous);
+
+  if (reduction == Reduction::None) {
+    const int64_t batch_size = input.size(0);
+    const int64_t H = input.size(2);
+    const int64_t W = input.size(3);
+
+    output.resize_({batch_size, H, W});
+    auto input_acc = input.accessor<scalar_t, 4>();
+    auto output_acc = output.accessor<scalar_t, 3>();
+    auto target_acc = target.accessor<int64_t, 3>();
+
+    // we check target indicies but cannot throw inside parallel_for
```

Comment:
nit: indices. Also, I thought `parallel_for` does propagate exceptions correctly; at least, the OpenMP implementation implies that it does error handling:

```
#ifdef _OPENMP
  std::atomic_flag err_flag = ATOMIC_FLAG_INIT;
  std::exception_ptr eptr;
  // choose number of tasks based on grain size and number of threads
  int64_t num_threads = omp_in_parallel() ? 1 : omp_get_max_threads();
  if (grain_size > 0) {
    num_threads = std::min(num_threads, divup((end - begin), grain_size));
  }

#pragma omp parallel num_threads(num_threads)
  {
    int64_t num_threads = omp_get_num_threads();
    int64_t tid = omp_get_thread_num();
    int64_t chunk_size = divup((end - begin), num_threads);
    int64_t begin_tid = begin + tid * chunk_size;
    if (begin_tid < end) {
      try {
        f(begin_tid, std::min(end, chunk_size + begin_tid));
      } catch (...) {
        if (!err_flag.test_and_set()) {
          eptr = std::current_exception();
        }
      } 
    } 
  }     
  if (eptr) {
    std::rethrow_exception(eptr);
  }
#else
  f(begin, end);
#endif
```

Reply:
@ezyang Worker-thread exception propagation is definitely a fine thing to have - sry, I did not check the parallel_for internals! The code in the PR was a modification of the old TH `nll_loss` implementation which I then also used in the 2d version. The range check could and maybe should be implemented with regular `TORCH_CHECK()` inside the `parallel_for` lambda-body.
General question: When the facebook-github-bot 'landing' notification has been posted .. does that mean I must no longer push modifications to this PR but rather wait for the merge to happen and then later create a new PR to address further changes?

Reply:
It's OK to push new changes, that will just terminate the land and I'll try again later. fbcode master has been sickly so I've been rebasing.